feat: new code analysis page (#5603)
JetBrains/kotlin-web-site: docs/topics/jvm/jvm-code-analysis.md

   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```bash
   ./gradlew ktlintFormat
   ```

   </tab>
   </tabs>

For more information on available features and rules, see the [ktlint documentation](https://pinterest.github.io/ktlint/latest/).

## Code analysis with detekt

[detekt](https://github.com/detekt/detekt) is a static code analysis tool for Kotlin that detects code smells, complexity issues,
and potential bugs.

To integrate detekt into your project:

1. Add the plugin to your build file:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```xml
   <!-- pom.xml -->
   <plugin>
       <groupId>com.github.ozsie</groupId>
       <artifactId>detekt-maven-plugin</artifactId>
       <version>1.23.8</version>
       <executions>
           <execution>
               <phase>verify</phase>
               <goals>
                   <goal>check</goal>
               </goals>
           </execution>
       </executions>
   </plugin>
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">
   
   ```kotlin
   // build.gradle.kts
   plugins {
       id("io.gitlab.arturbosch.detekt") version "1.23.8"
   }
   ```

   </tab>
   </tabs>

> 2. (Optional) You can also add a `detekt.yml` configuration file to the root of your project to customize the rules:

Reviewer: detekt does not auto-discover a root config (verified against detekt's Gradle getting-started docs). Without detekt { config.setFrom("detekt.yml") } in Gradle (or the <config> parameter for the com.github.ozsie Maven plugin)

## Code quality with SonarSource

[SonarQube](https://www.sonarsource.com/products/sonarqube/) and [SonarCloud](https://www.sonarsource.com/products/sonarcloud/)
from SonarSource provide deep static analysis for Kotlin projects, including bug detection, vulnerability scanning,
and code coverage tracking through a web dashboard.

To analyze your project with SonarQube:

1. Add the plugin to your build file:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```xml
   <!-- pom.xml -->
   <plugin>
       <groupId>org.sonarsource.scanner.maven</groupId>
       <artifactId>sonar-maven-plugin</artifactId>
       <version>5.7.0.6970</version>
   </plugin>
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```kotlin
   // build.gradle.kts
   plugins {
       id("org.sonarqube") version "6.2.0.5505"
   }
   ```

   </tab>
   </tabs>

2. (Optional) You can also add a `sonar-project.properties` file to the root of your project to customize quality gate
   conditions. For example, to fail the build if code coverage drops below 80% or if new bugs are introduced:

   ```properties
   sonar.qualitygate.wait=true

>    sonar.coverage.minimum=80

Reviewer: can't find it here: https://docs.sonarsource.com/sonarqube-server/analyzing-source-code/scanners/sonarscanner-for-gradle

and in my repro they doesn't work. could you please check?

[//]: # (title: Code quality tools in Kotlin projects)

Code quality tools help you enforce coding standards, detect bugs early, measure test coverage, and maintain code quality
across your Kotlin projects. This guide shows how to integrate popular tools like [ktlint](https://pinterest.github.io/ktlint/),

> [detekt](https://detekt.dev/), [SonarSource](https://www.sonarsource.com/), and [Kover](https://kotlin.github.io/kotlinx-kover/maven-plugin/)

Reviewer 2: SonarSource is a company, not the tool, right? I would mention SonarQube here and later, with SonarCloud as an alternative.

Author: Right, we still need to mention the company name for discoverability

Reviewer 2: ~~I think like with links, it would maybe make sense to mention the company at the top and the tool at the bottom.~~ I noticed that you did it the other way around, that works too 👍 

But my main issue here is with the phrasing: "popular tools like ... SonarSource" doesn't work, because SonarSource isn't a tool; it reads like a typo in a way.

> [detekt](https://detekt.dev/), [SonarSource](https://www.sonarsource.com/), and [Kover](https://kotlin.github.io/kotlinx-kover/maven-plugin/)

Reviewer 2: ```suggestion
[detekt](https://detekt.dev/), [SonarSource](https://www.sonarsource.com/), and [Kover](https://github.com/Kotlin/kotlinx-kover)
```
Would this be a better link than the Maven plugin page?

into your Maven or Gradle build.

## Code formatting with ktlint

> [ktlint](https://pinterest.github.io/ktlint/) is a Kotlin linter and formatter that enforces the

Reviewer 2: I think this is supposed to be the link? https://ktlint.github.io/ktlint/

They also mention supporting the [Android style guide](https://developer.android.com/kotlin/style-guide) by default which can be an additional draw for people because it's noticeably different from Kotlin conventions.

Author: So I reviewed all the links, I'd like to keep links to the tools' main websites in the intro, and refer to github repos in subsections

official Kotlin [coding conventions](coding-conventions.md) with no extra configuration.

ktlint checks rules such as indentation, spacing around operators, import ordering, and trailing commas. If a violation

> is found, the build fails with a message indicating the file and line number.

Reviewer 2: Not to misrepresent ktlint — it advertises not only finding but fixing straightforwardly fixable violations.

           </execution>
       </executions>
   </plugin>
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```kotlin
   // build.gradle.kts
   plugins {
       id("org.jlleitschuh.gradle.ktlint") version "12.3.0"
   }
   ```

   </tab>
   </tabs>

2. Run the linter to check your code style:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```bash
   mvn ktlint:check
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```bash
   ./gradlew ktlintCheck
   ```

   </tab>
   </tabs>

3. (Optional) You can also add an `.editorconfig` file to the root of your project to customize rules. For example, to
   allow wildcard imports and disable trailing comma enforcement:

   ```ini
   [*.{kt,kts}]
   ij_kotlin_imports_layout = *
   ktlint_standard_trailing-comma-on-call-site = disabled
   ktlint_standard_trailing-comma-on-declaration-site = disabled
   ```

4. To automatically fix formatting issues, run:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```bash
   mvn ktlint:format
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```bash
   ./gradlew ktlintFormat
   ```

   </tab>
   </tabs>

> For more information on available features and rules, see the [ktlint documentation](https://pinterest.github.io/ktlint/latest/).

Reviewer 2: This link also redirects to Pinterest careers instead of the ktlint GitHub.

## Code analysis with detekt

> [detekt](https://github.com/detekt/detekt) is a static code analysis tool for Kotlin that detects code smells, complexity issues,

Reviewer 2: I'm not a real developer, but this is the first time I hear the term "code smell" — I would either explain it here, or change for something less slangy.

Author: Yeah, it's the first feature on their list, they also have a github tag to find it

Reviewer 2: I understand, but their feature list can also use a definition there. Because when I have to google a term used in the promotional material, that's not a very good promotion. But again, if the target audience is supposed to be fine with it, I'm not pushing here :)

> [detekt](https://github.com/detekt/detekt) is a static code analysis tool for Kotlin that detects code smells, complexity issues,

Reviewer 2: ```suggestion
[detekt](https://detekt.dev/) is a static code analysis tool for Kotlin that detects code smells, complexity issues,
```
To align with the link given at the beginning.

   ```

4. Reference the configuration file in your build file so that detekt can apply the new rules:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```xml
   <!-- pom.xml -->
   <plugin>
       <groupId>com.github.ozsie</groupId>
       <artifactId>detekt-maven-plugin</artifactId>
       <version>1.23.8</version>
       <configuration>
           <config>config/detekt/detekt.yml</config>
       </configuration>
       <executions>
           <execution>
               <phase>verify</phase>
               <goals>
                   <goal>check</goal>
               </goals>
           </execution>
       </executions>
   </plugin>
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```kotlin
   // build.gradle.kts
   detekt {
       toolVersion = "1.23.8"
       config.setFrom(file("config/detekt/detekt.yml"))
       buildUponDefaultConfig = true
   }
   ```

   </tab>
   </tabs>

5. Run the analysis:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```bash
   mvn detekt:check
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```bash
   ./gradlew detekt
   ```

   </tab>
   </tabs>

> detekt will produce a report listing all rule violations with their severity, file location, and a description of the issue.

Reviewer 2: ```suggestion
detekt produces a report listing all rule violations with their severity, file location, and a description of the issue.
```

For more information, see the detekt documentation for [Gradle](https://detekt.dev/docs/intro) and [Maven](https://github.com/Ozsie/detekt-maven-plugin).

## Code quality with SonarSource

[SonarQube](https://www.sonarsource.com/products/sonarqube/) and [SonarCloud](https://www.sonarsource.com/products/sonarcloud/)
from SonarSource provide deep static analysis for Kotlin projects, including bug detection, vulnerability scanning,
and code coverage tracking through a web dashboard.

To analyze your project with SonarQube:

1. Add the plugin to your build file:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```xml
   <!-- pom.xml -->
   <plugin>
       <groupId>org.sonarsource.scanner.maven</groupId>
       <artifactId>sonar-maven-plugin</artifactId>
       <version>5.7.0.6970</version>
   </plugin>
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```kotlin
   // build.gradle.kts
   plugins {
       id("org.sonarqube") version "6.2.0.5505"
   }

   sonar {
       properties {
           property("sonar.projectKey", "my-project")
           property("sonar.host.url", "http://localhost:9000")
       }
   }
   ```

   </tab>
   </tabs>

> 2. (Optional) Configure analysis properties in your build file. For example, to make the build wait for the

Reviewer 2: ```suggestion
2. (Optional) Configure analysis properties. For example, to make the build wait for the
```

Maven instructions don't mention a build file, so I assume this has to be more abstract.

   > or SonarCloud web interface under **Quality Gates**, not in the build file.
   >
   {style="note"}

3. Run the analysis against your SonarQube server:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```bash
   mvn verify sonar:sonar \
     -Dsonar.projectKey=my-project \
     -Dsonar.host.url=http://localhost:9000 \
     -Dsonar.token=YOUR_TOKEN
   ```

   For SonarCloud, replace the host URL with `https://sonarcloud.io` and provide your organization key:

   ```bash
   mvn verify sonar:sonar \
     -Dsonar.projectKey=my-project \
     -Dsonar.organization=my-org \
     -Dsonar.host.url=https://sonarcloud.io \
     -Dsonar.token=YOUR_TOKEN
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   To run the analysis, use the `sonar` task and provide your authentication token:

   ```bash
   ./gradlew sonar \
     -Dsonar.token=YOUR_TOKEN
   ```

   By default, the analysis is run against a local SonarQube server. To use SonarCloud, update the `sonar {}` block
   in your `build.gradle.kts` to use `https://sonarcloud.io` and add your organization key:

   ```kotlin
   sonar {
       properties {

>            property("sonar.projectKey", "my-project")

Reviewer 2: ```suggestion
           property("sonar.projectKey", "example-project")
```

>            property("sonar.organization", "my-org")

Reviewer 2: ```suggestion
           property("sonar.organization", "example-org")
```
The nittest nit.

           property("sonar.host.url", "https://sonarcloud.io")
       }
   }
   ```

   </tab>
   </tabs>

4. Open the SonarQube or SonarCloud dashboard to review the results. The dashboard shows issues grouped by type
   (bug, vulnerability, code smell) and severity.

For more information, see the [SonarSource documentation](https://docs.sonarsource.com/).

## Code coverage with Kover

> [Kover](https://kotlin.github.io/kotlinx-kover/) is the official JetBrains code coverage tool for Kotlin.

Reviewer 2: Here also I would link to a more informative repo README https://github.com/Kotlin/kotlinx-kover

It measures which lines and branches of your code are covered by tests and generates human-readable reports.

> Unlike JaCoCo, Kover understands Kotlin-specific constructs such as inline functions and data classes, so it reports

Reviewer 2: ```suggestion
Unlike [JaCoCo](https://github.com/jacoco/jacoco), Kover correctly interprets Kotlin-specific constructs such as inline functions and data classes, so it reports
```

To avoid anthropomorphism.

               <goals>
                   <goal>instrumentation</goal>
               </goals>
           </execution>
           <execution>
               <id>kover-verify</id>
               <goals>
                  <goal>verify</goal>
               </goals>
           </execution>
           <execution>
               <id>kover-report-xml</id>
               <goals>
                   <goal>report-xml</goal>
               </goals>
           </execution>
           <execution>
               <id>kover-report-html</id>
               <goals>
                   <goal>report-html</goal>
               </goals>
           </execution>
       </executions>
   </plugin>
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```kotlin
   // build.gradle.kts
   plugins {
       id("org.jetbrains.kotlinx.kover") version "0.9.8"
   }
   ```

   </tab>
   </tabs>

2. Run your tests to collect coverage data and generate the report:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```bash
   mvn verify
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```bash
   ./gradlew koverHtmlReport
   ```

   </tab>
   </tabs>

> 3. Open the HTML report  generated in the `target/kover/html/` directory (`build/reports/kover/html/` for Gradle) to

Reviewer 2: ```suggestion
3. Open the HTML report generated in the `target/kover/html/` directory (`build/reports/kover/html/` for Gradle) to
```

   review line-by-line coverage.
4. (Optional) To enforce a minimum coverage threshold that fails the build if the conditions are not met, you can add
   a coverage verification configuration to your build file. For example:

   <tabs group="build-system">
   <tab title="Maven" group-key="maven">

   ```xml
   <!-- pom.xml -->
   <configuration>
     <!-- Create new coverage verification rule -->
     <rules>
         <rule>
             <bounds>
                 <bound>

>                      <minValue>80</minValue>

Reviewer 2: ```suggestion
                     <minValue>50</minValue>
```
To align it with the Gradle example?

                     <coverageUnits>LINE</coverageUnits>
                     <aggregationForGroup>COVERED_PERCENTAGE</aggregationForGroup>
                 </bound>
             </bounds>
         </rule>
     </rules>
   </configuration>
   ```

   </tab>
   <tab title="Gradle" group-key="gradle">

   ```kotlin
   // build.gradle.kts
   // Add a `kover` configuration block
   import kotlinx.kover.gradle.plugin.dsl.*
   
   kover {
       reports {
           total {
               log {
                   aggregationForGroup = AggregationType.COVERED_PERCENTAGE
                   coverageUnits = CoverageUnit.LINE
               }
               verify {
                   rule {

>                        minBound(50)

Reviewer 2: I ran this in a project with 0% coverage in the report, but the build did not fail :(
It was a KMP project, but there are no tests anywhere, so I expected this to fail regardless.

Author: the verification command for gradle was wrong, thanks

                   }
               }
           }
       }
   }
   ```

   </tab>
   </tabs>

For more information on configuring verification rules, see the Kover documentation for [Maven](https://kotlin.github.io/kotlinx-kover/maven-plugin/)
and [Gradle](https://kotlin.github.io/kotlinx-kover/gradle-plugin/).

## Other tools

> Besides ktlint, detekt, SonarSource, and Kover, you can use other tools to improve Kotlin code quality:

Reviewer 2: ```suggestion
Besides ktlint, detekt, SonarQube, and Kover, you can use other tools to improve Kotlin code quality:
```

[//]: # (title: Code quality tools in Kotlin projects)

Code quality tools help you enforce coding standards, detect bugs early, measure test coverage, and maintain code quality

> across your Kotlin projects. This guide focusesand shows how to integrate popular tools like

Reviewer 2: ```suggestion
across your Kotlin projects. This guide focuses and shows how to integrate popular tools like
```
But I would choose either "focuses on" or "shows how" because together they don't work very well to me.
